### ISSUE TITLE
Potentially printing provenance more than once in LoadErrors

### ISSUE THREAD

**Issue by Cynical-Optimist** [buildstream-migration/bst-staging#142]
[See original issue on GitLab](https://gitlab.com/BuildStream/buildstream/-/issues/142)
In GitLab by [[Gitlab user @tristanvb]](https://gitlab.com/tristanvb) on Oct 31, 2017, 07:07

As discovered in MR !129, it is possible that by way of resolving option expressions, we end up prepending provenance to errors which already have provenance, resulting in errors such as:

```
element.bst [line 11 column 4]: element.bst [line 11 column 4]: Oh crap something went wrong here
```

This is due to how `_options/optionpool.py` prepends provenance to error messages. As described in [this comment](https://gitlab.com/BuildStream/buildstream/merge_requests/129/diffs#note_45173680), that prepending is intended to add provenance only when a `LoadError` is raised from `self.evaluate()`, the `_yaml.composite()` line should be moved out of the try / except block so as to avoid this.

However, it's not entirely that simple, making this change will also cause the error message introduced by MR !129 to regress and lack provenance information, so both of these must be fixed together.

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tristanvb]](https://gitlab.com/tristanvb) on Oct 31, 2017, 07:08

mentioned in merge request !129

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tlater]](https://gitlab.com/tlater) on Nov 2, 2017, 17:07

created branch [`142-potentially-printing-provenance-more-than-once-in-loaderrors`](https://gitlab.com/BuildStream/buildstream/compare/master...142-potentially-printing-provenance-more-than-once-in-loaderrors)

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tlater]](https://gitlab.com/tlater) on Nov 2, 2017, 17:08

mentioned in commit 7d8b0a95cca0391d998f5853da79d766332c393f

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tlater]](https://gitlab.com/tlater) on Nov 23, 2017, 14:47

mentioned in commit cd06e0ecf63fe7d83457f482b6e782ccd3588dbe

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tlater]](https://gitlab.com/tlater) on Nov 29, 2017, 11:09

mentioned in commit 36354bb9df300afb4fdd4e937ebad49ec3a40e2e

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tlater]](https://gitlab.com/tlater) on Nov 29, 2017, 12:03

mentioned in commit 00ee89f85446105ce2e995442900c11231f86b4d

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tlater]](https://gitlab.com/tlater) on Nov 29, 2017, 14:58

mentioned in commit f67521ab6563a32e2914ff4e454c8c9bc6d16e37

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tlater]](https://gitlab.com/tlater) on Nov 29, 2017, 15:07

mentioned in commit 0915e5a43215c4618030b7b62d88af441e76cbe4

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tlater]](https://gitlab.com/tlater) on Nov 29, 2017, 15:15

mentioned in merge request !170

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tlater]](https://gitlab.com/tlater) on Nov 29, 2017, 17:05

mentioned in commit df59ecf71c203b354d44efa6c25e71c4f2a9c648

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tristanvb]](https://gitlab.com/tristanvb) on Nov 30, 2017, 17:46

mentioned in commit af5157422f1f24617936c90a370e58c45ae62647

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tristanvb]](https://gitlab.com/tristanvb) on Nov 30, 2017, 17:53

mentioned in commit 3f559f2a295fd903f72e553485d781e01a1784e6

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tristanvb]](https://gitlab.com/tristanvb) on Nov 30, 2017, 17:54

This is fixed with commit 3f559f2a

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tristanvb]](https://gitlab.com/tristanvb) on Nov 30, 2017, 17:54

closed
